### PR TITLE
Neumann boundary conditions and differentiation between the pressure's and phi's affine constraints.

### PR DESCRIPTION
**Pull request by j507**
The pull-request is mainly to start the review process. I tested the Neumann boundary conditions with the Couette flow as you recommended. Results are those expected. There are two things I would like to add/check before merging:

- [x] Convergece test for the Guermond problem 10.3 ~Include the time dependent term of the Couette flow solution and do some ,convergence tests.~
- [x] **Update: See comment below**. The implementation of `phi`'s `AffineConstraint` seems to introduce a small variation in the solution. I only noticed it when running `step-35` with adaptive meshing (50 time steps and adaptive meshing every 20th step). Before the second adaptive meshing operation, there is no difference to be seen in the terminal output when its decimal precision is 6. When one sets it to 10 there are small differences to be seen. These variations apparently become noticed by the estimated error used during the adaptive meshing procedure, as the `master` produces

```
Preparing coarsening and refining...
   Number of cells set for refinement: 110
   Number of cells set for coarsening: 3147
Executing coarsening and refining...
   Number of global active cells:      1929
   Number of cells on each (level):    0 (0), 0 (1), 747 (2), 742 (3), 440 (4)  


Preparing coarsening and refining...
   Number of cells set for refinement: 70
   Number of cells set for coarsening: 1092
Executing coarsening and refining...
   Number of global active cells:      1476
   Number of cells on each (level):    0 (0), 120 (1), 298 (2), 634 (3), 360 (4), 64 (5)  

Step =     50, Time = 2.500000e-01, Time step = +0.000000e+00 Velocity = (+1.391462e+00, +9.407433e-02) Pressure = +1.189951e+00
```

Screenshot from 2020-11-27 01-29-37

while the `neumann_boundary_condition` branch produces

```
Preparing coarsening and refining...
   Number of cells set for refinement: 110
   Number of cells set for coarsening: 3146
Executing coarsening and refining...
   Number of global active cells:      1929
   Number of cells on each (level):    0 (0), 0 (1), 747 (2), 742 (3), 440 (4) 

Preparing coarsening and refining...
   Number of cells set for refinement: 71
   Number of cells set for coarsening: 1093
Executing coarsening and refining...
   Number of global active cells:      1482
   Number of cells on each (level):    0 (0), 120 (1), 297 (2), 637 (3), 364 (4), 64 (5) 

Step =     50, Time = 2.500000e-01, Time step = +0.000000e+00 Velocity = (+1.391414e+00, +9.409932e-02) Pressure = +1.189992e+00
```
Screenshot from 2020-11-27 01-27-24

From a visual inspection the triangulation difference is located just near the square obstacle, which would explain the difference at the last time step. Apart from very small variances in the lift coefficient (which seems to be just really sensitive) by the `DFG` problem, the terminal output from all problems including the other values of the `DFG` problem do not show any variance at all.

This is to be merged **after** #46.

### CODE REVIEW

**Review by sebglane — APPROVED** [review]
I added some comments. The one with synchronization would be worthwhile to check. The others are more or less simple minor improvements.

Good to go.

File: applications/ThermalTGV.cc [hunk 1]
```diff
@@ -260,10 +260,6 @@ void ThermalTGV<dim>::solve(const unsigned int &level)
   velocity->reinit();
   error.reinit(temperature->solution);
   initialize();
-
-  // Advances the time to t^{k-1}, either t^0 or t^1
-  for (unsigned int k = 1; k < time_stepping.get_order(); ++k)
-    time_stepping.advance_time();
```
Comment (j507) [inline applications/ThermalTGV.cc:266]: @sebglane  The removal of these lines is what causes the convergence rate to change in this branch. If added again one gets:
```
                               Temperature convergence table
==============================================================================================
level    dt    cells dofs    hmax           L2                 H1               Linfty       
    8 1.00e-01 65536 66049 5.52e-03 3.422045e-04     - 1.459549e-02     - 6.387538e-04     - 
    8 5.00e-02 65536 66049 5.52e-03 9.723780e-05 -1.82 1.431300e-02 -0.03 1.483619e-04 -2.11 
    8 2.50e-02 65536 66049 5.52e-03 3.221458e-05 -1.59 1.429360e-02 -0.00 1.721641e-05 -3.11 
    8 1.25e-02 65536 66049 5.52e-03 1.626306e-05 -0.99 1.429249e-02 -0.00 1.647540e-05 -0.06 
    8 6.25e-03 65536 66049 5.52e-03 1.262168e-05 -0.37 1.429246e-02 -0.00 2.492815e-05  0.60 
    8 3.12e-03 65536 66049 5.52e-03 1.178615e-05 -0.10 1.429247e-02  0.00 2.696698e-05  0.11 
    8 1.56e-03 65536 66049 5.52e-03 1.157736e-05 -0.03 1.429247e-02  0.00 2.749647e-05  0.03 
    8 7.81e-04 65536 66049 5.52e-03 1.154823e-05 -0.00 1.429247e-02  0.00 2.754702e-05  0.00 
```
i.e. the results from `mit_benchmark`. One can then assume that the error in `master` is not caused by the `boundary_conditions.*` files.
